EIPS/eip-5450.md

```diff
@@ -58,48 +58,50 @@ In the second validation phase control-flow analysis is performed on the code.
    1. The next instruction for all instructions other than terminating instructions and unconditional jump.
    2. All targets of a conditional or unconditional jump.
-3. For each successor instruction:
+4. For each successor instruction:
    1. **Check** if the instruction is present in the code (i.e. execution must not "fall off" the code).
-   2. If the instruction does not have stack height recorded (visited for the first time):
-      1. Record the instruction stack height as the value computed in 1.2.
-      2. Add the instruction to the *worklist*.
-   3. Otherwise, **check** if the recorded stack height equals the value computed in 1.2.
+   2. If the successor is reached via forwards jump or sequential flow from previous instruction:
+      1. If the instruction does not have stack height bounds recorded (visited for the first time), record the instruction stack height bound as the value computed in 2.2.
+      2. Otherwise instruction was already visited (by previously seen forward jump). Update this instruction's recorded stack height bounds so that they contain the bounds computed in 2.2, i.e. `target_stack_min = min(target_stack_min, current_stack_min)` and `target_stack_max = max(target_stack_max, current_stack_min)`, where `(target_stack_min, target_stack_max)` are successor bounds and `(current_stack_min, current_stack_max)` are bounds computed in 2.2.
+   3. If the successor is reached via backwards jump, **check** if the recorded stack height bounds equal the value computed in 2.2. Validation fails if they are not equal, i.e. we see backwards jump to a different stack height.
 
-When *worklist* is empty:
-
-1. **Check** if all instruction were reached by the analysis.
-2. Determine the function maximum operand stack height:
-   1. Compute the maximum stack height as the maximum of all recorded stack heights.
+After all instructions are visited, determine the function maximum operand stack height:
+   1. Compute the maximum stack height as the maximum of all recorded stack heights upper bounds.
    2. **Check** if the maximum stack height does not exceed the limit of 1023 (`0x3FF`).
    3. **Check** if the maximum stack height matches the corresponding code section's `max_stack_height` within the type section.
 
 The computational and space complexity of this pass is *O(len(code))*. Each instruction is visited at most once.
 
 ### Execution
 
-Given the deploy-time validation guarantees, an EVM implementation is not required anymore to have run-time stack underflow nor overflow checks for each executed instruction. The exception is the `CALLF` performing operand stack overflow check for the entire called function.
+Given the deploy-time validation guarantees, an EVM implementation is not required anymore to have run-time stack underflow nor overflow checks for each executed instruction. The exception is the `CALLF` and `JUMPF` performing operand stack overflow check for the entire called function.
 
 ## Rationale
```

Comment:
Once we're done with the specifications part, needs adding justification of going down the relaxed validation path

Reply:
Added in Rationale.